### PR TITLE
sync/task_pool: addTaskAlways

### PR DESCRIPTION
**Pull request by micln**
**What this PR does**:

- new methods : addTaskAlways addTaskBalance
- add testcase and benchmark

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: sync/task_pool.go [hunk 1]
```diff
@@ -33,9 +33,9 @@ const (
 /////////////////////////////////////////
 
 type TaskPoolOptions struct {
-	tQLen      int // task queue length
-	tQNumber   int // task queue number
-	tQPoolSize int // task pool size
+	tQLen           int // task queue length
+	tQSizePerWorker int // task queue number
```
Comment (AlexStocks) [inline sync/task_pool.go]: pls do not change this member name.

**Comment by wongoo** [comment]
Should provide:
- What this PR does
- Special notes for your reviewer

File: sync/worker.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+func NewWorker(bufferSize int64) Worker {
+	return &GoWorker{
+		ChannelTaskQueue: NewChannelTaskQueue(bufferSize),
+		sigStop:          nil,
+	}
+}
+
+type ChannelWorkerManager struct {
+	TaskQueue // todo 这个地方可能是瓶颈
```
Comment (AlexStocks) [inline sync/worker.go]: pls do not use Chinese remark in our project.

File: sync/worker.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+func (c *ChannelWorkerManager) dispatch() {
+	go func() {
+		for {
+			var t Task
+			select {
+			case t = <-c.Pop():
+			}
+
+			// todo 选一个闲置的 worker
```
Comment (AlexStocks) [inline sync/worker.go]: translate it into english.

File: sync/task_pool_test.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+	"runtime"
+	"sync/atomic"
+	"testing"
+)
+
+func BenchmarkTaskPool_AddTask(b *testing.B) {
+	tp := NewTaskPool(
+		WithTaskPoolTaskPoolSize(runtime.NumCPU()),
+		WithTaskPoolTaskQueueNumber(runtime.NumCPU()),
+		// WithTaskPoolTaskQueueLength(1),
```
Comment (zouyx) [inline sync/task_pool_test.go]: if useless then delete

File: sync/task_pool_test.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+	)
+
+	var cnt int64 = 0
+	task := func() {
+		atomic.AddInt64(&cnt, 1)
+		// time.Sleep(1 * time.Microsecond)
```
Comment (zouyx) [inline sync/task_pool_test.go]: if useless then delete

File: sync/task_pool_test.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	}
+
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			tp.AddTask(task)
+		}
+	})
+	// tp.Close()
```
Comment (zouyx) [inline sync/task_pool_test.go]: if useless then delete

File: sync/task_pool_test.go [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+				tp.AddTask(task)
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+	tp.Close()
+
+	if taskCnt != *cnt {
+		t.Error("want ", taskCnt, " got ", *cnt)
```
Comment (micln) [inline sync/task_pool_test.go]: @AlexStocks this assert failed sometimes

image

Comment (micln) [reply sync/task_pool_test.go]: maybe the pool receive done signal before than send channel in 

image

Comment (micln) [reply sync/task_pool_test.go]: can you help me check if the test code has errors  ?

File: sync/task_pool_test.go (hunk 5 above)
Comment (micln) [reply sync/task_pool_test.go]: thx

File: sync/task_pool.go [hunk 8]
```diff
@@ -158,14 +159,54 @@ func (p *TaskPool) run(id int, q chan task) error {
...
+func (p *TaskPool) AddTaskAlways(t task) {
 	id := atomic.AddUint32(&p.idx, 1) % uint32(p.tQNumber)
 
 	select {
 	case <-p.done:
 		return
 	case p.qArray[id] <- t:
+		return
+	default:
+		go t()
```
Comment (AlexStocks) [inline sync/task_pool.go]: pls do not use 'go'. use 'gxruntime.GoSafely' instead.

File: sync/task_pool.go [hunk 9]
```diff
@@ -158,14 +159,54 @@ func (p *TaskPool) run(id int, q chan task) error {
...
+
+	// try len/2 times to lookup idea queue
+	for i := 0; i*2 < length; i++ {
+		i := rand.Intn(length)
+		if p.tryAddTaskTo(t, i) {
+			return
+		}
+	}
+
+	go t()
```
Comment (AlexStocks) [inline sync/task_pool.go]: use 'gxruntime.GoSafely' instead.

File: sync/task_pool.go [hunk 10]
```diff
@@ -158,14 +159,54 @@ func (p *TaskPool) run(id int, q chan task) error {
...
+		go t()
+	}
+}
+
+// AddTaskBalance add task to idea queue
+// do it immediately when no idea queue
+func (p *TaskPool) AddTaskBalance(t task) {
+	length := len(p.qArray)
+
+	// try len/2 times to lookup idea queue
```
Comment (AlexStocks) [inline sync/task_pool.go]: hey, guy. I do not know what "idea" meas. do u mean "idle"?

File: sync/task_pool.go [hunk 11]
```diff
@@ -158,14 +159,54 @@ func (p *TaskPool) run(id int, q chan task) error {
...
+
+	select {
+	case <-p.done:
+		return
+	case p.qArray[id] <- t:
+	}
+}
+
+// AddTaskAlways add task to queues or do it immediately
+func (p *TaskPool) AddTaskAlways(t task) {
```
Comment (wongoo) [inline sync/task_pool.go]: suggest to rename func name to `ImmediateAddTask`

File: sync/task_pool.go [hunk 12]
```diff
@@ -158,14 +159,54 @@ func (p *TaskPool) run(id int, q chan task) error {
...
+	}
+}
+
+// AddTaskBalance add task to idea queue
+// do it immediately when no idea queue
+func (p *TaskPool) AddTaskBalance(t task) {
+	length := len(p.qArray)
+
+	// try len/2 times to lookup idea queue
+	for i := 0; i*2 < length; i++ {
```
Comment (wongoo) [inline sync/task_pool.go]: why not `i < length/2`?

File: sync/task_pool.go [hunk 13]
```diff
@@ -158,14 +159,54 @@ func (p *TaskPool) run(id int, q chan task) error {
...
+		i := rand.Intn(length)
```
Comment (wongoo) [inline sync/task_pool.go]: use variable `i` again, change it to another

File: sync/task_pool.go [hunk 14]
```diff
@@ -158,14 +159,54 @@ func (p *TaskPool) run(id int, q chan task) error {
...
-// add task
+// AddTask wait idle worker add task
 func (p *TaskPool) AddTask(t task) {
+	idx := atomic.AddUint32(&p.idx, 1)
+	id := idx % uint32(p.tQNumber)
+
+	select {
+	case <-p.done:
+		return
+	case p.qArray[id] <- t:
```
Comment (wongoo) [inline sync/task_pool.go]: I think `AddTask` should make sure task being added and executed, so change it to :
```go
select {
	case <-p.done:
		return
       default:
                p.qArray[id] <- t
}
```

File: sync/task_pool.go [hunk 15]
```diff
@@ -158,14 +159,54 @@ func (p *TaskPool) run(id int, q chan task) error {
...
 	case p.qArray[id] <- t:
+		return
+	default:
+		go t()
+	}
+}
+
+// AddTaskBalance add task to idea queue
+// do it immediately when no idea queue
+func (p *TaskPool) AddTaskBalance(t task) {
```
Comment (wongoo) [inline sync/task_pool.go]: suggest to rename func name to `BalanceAddTask`

File: sync/task_pool.go [hunk 16]
```diff
@@ -158,14 +159,54 @@ func (p *TaskPool) run(id int, q chan task) error {
...
+	length := len(p.qArray)
+
+	// try len/2 times to lookup idea queue
+	for i := 0; i*2 < length; i++ {
+		i := rand.Intn(length)
+		if p.tryAddTaskTo(t, i) {
```
Comment (wongoo) [inline sync/task_pool.go]: how about:
```go
select {
	case p.qArray[rand.Intn(length)] <- t:
		return
	default:
		continue
	}
```

File: runtime/goroutine.go [hunk 17]
```diff
@@ -26,7 +26,7 @@ import (
 )
 
 // GoSafely wraps a `go func()` with recover()
-func GoSafely(wg *sync.WaitGroup, ignoreRecover bool, handler func(), finalFunc func(r interface{})) {
+func GoSafely(wg *sync.WaitGroup, ignoreRecover bool, handler func(), catchFunc func(r interface{})) {
```
Comment (micln) [inline runtime/goroutine.go]: this func only called when catch exception, seems like to `catch (e) {}` not `finally {}`

File: sync/task_pool.go [hunk 18]
```diff
@@ -127,10 +131,24 @@ func (p *TaskPool) start() {
 		p.wg.Add(1)
 		workerID := i
 		q := p.qArray[workerID%p.tQNumber]
-		go p.run(int(workerID), q)
+		p.safeRun(workerID, q)
```
Comment (micln) [inline sync/task_pool.go]: pls review this func

File: runtime/goroutine.go (hunk 17 above)
Comment (micln) [inline runtime/goroutine.go]: this func only called when catch exception, seems like to `catch (e) {}` not `finally {}`

File: sync/task_pool.go (hunk 11 above)
Comment (micln) [reply sync/task_pool.go]: how about `AddTaskImmediately`? same prefix is friendly for find and  IDE's auto-complete

File: sync/task_pool.go (hunk 10 above)
Comment (micln) [reply sync/task_pool.go]: sorry

File: sync/task_pool.go (hunk 14 above)
Comment (micln) [reply sync/task_pool.go]: if p.done is ok, both `case` and `default` are not work.

I add a return value means success status, both them could get a false value

File: sync/task_pool_test.go (hunk 7 above)
Comment (micln) [reply sync/task_pool_test.go]: got the reason by the log https://github.com/dubbogo/gost/pull/25/files#diff-e3fb3a64c85404e71d3fd89a55036c36R144

File: sync/task_pool.go (hunk 8 above)
Comment (micln) [reply sync/task_pool.go]: updated.

and please review func p.safeRun and goSafely https://github.com/dubbogo/gost/pull/25/files#diff-e3fb3a64c85404e71d3fd89a55036c36R134

File: sync/task_pool.go (hunk 9 above)
Comment (micln) [reply sync/task_pool.go]: updated

File: sync/task_pool.go (hunk 13 above)
Comment (micln) [reply sync/task_pool.go]: fixed

File: sync/task_pool.go (hunk 12 above)
Comment (micln) [reply sync/task_pool.go]: updated

File: sync/task_pool.go (hunk 16 above)
Comment (micln) [reply sync/task_pool.go]: updted

File: runtime/goroutine.go (hunk 17 above)
Comment (AlexStocks) [reply runtime/goroutine.go]: ok

File: sync/task_pool.go [hunk 19]
```diff
@@ -19,8 +19,12 @@ package gxsync
 
 import (
 	"fmt"
+	"log"
+	"math/rand"
 	"sync"
 	"sync/atomic"
+
+	gxruntime "github.com/dubbogo/gost/runtime"
```
Comment (AlexStocks) [inline sync/task_pool.go]: move it to another import.

File: sync/task_pool.go [hunk 20]
```diff
@@ -127,10 +131,25 @@ func (p *TaskPool) start() {
...
+		func() {
+			err := p.run(int(workerID), q)
+			if err != nil {
+				// log error to stderr
+				log.Printf("gost/TaskPool.run error: %s", err.Error())
+			}
+		},
+		// catch handler
+		func(rec interface{}) {
+			p.safeRun(workerID, q)
```
Comment (wongoo) [inline sync/task_pool.go]: not exec again when panic, in my opinion

File: sync/task_pool.go (hunk 11 above)
Comment (wongoo) [reply sync/task_pool.go]: agree

File: sync/task_pool.go (hunk 20 above)
Comment (micln) [reply sync/task_pool.go]: it only start a new goroutine to consume the q, failed job won't repeat.

Comment (wongoo) [reply sync/task_pool.go]: call safeRun again in catch handler will exec again， or I misunderstand？

**Comment by wongoo** [comment]
some reviews are not resolved

File: sync/task_pool.go (hunk 20 above)
Comment (micln) [reply sync/task_pool.go]: the failed job won't exec agian. 
the method `safeRun` don't run job immediately, it only bind a queue to a goroutine.

Comment (micln) [reply sync/task_pool.go]: @AlexStocks dose I understand correctly ? 😄

Comment (wongoo) [reply sync/task_pool.go]: IMP，catch handler is not needed

Comment (AlexStocks) [reply sync/task_pool.go]: > IMP，catch handler is not needed

@micln what mr @wongoo said is right. If u use it in this way, it is like to gxruntime. GoUnterminated.

Comment (micln) [reply sync/task_pool.go]: removed

**Comment by AlexStocks** [comment]
LGTM

File: sync/task_pool.go [hunk 21]
```diff
@@ -158,15 +177,53 @@ func (p *TaskPool) run(id int, q chan task) error {
...
 }
 
-// add task
-func (p *TaskPool) AddTask(t task) {
-	id := atomic.AddUint32(&p.idx, 1) % uint32(p.tQNumber)
+// AddTask wait idle worker add task
+// return false when the pool is stop
+func (p *TaskPool) AddTask(t task) (ok bool) {
+	idx := atomic.AddUint32(&p.idx, 1)
+	id := idx % uint32(p.tQNumber)
```
Comment (wongoo) [inline sync/task_pool.go]: move the code into the default branch of select

File: sync/task_pool.go [hunk 22]
```diff
@@ -158,15 +177,53 @@ func (p *TaskPool) run(id int, q chan task) error {
...
+		p.qArray[id] <- t
+		return true
+	}
+}
+
+// AddTaskAlways add task to queues or do it immediately
+func (p *TaskPool) AddTaskAlways(t task) {
+	id := atomic.AddUint32(&p.idx, 1) % uint32(p.tQNumber)
+
+	select {
```
Comment (wongoo) [inline sync/task_pool.go]: is it need the check p.done?

File: sync/task_pool.go [hunk 23]
```diff
@@ -158,15 +177,53 @@ func (p *TaskPool) run(id int, q chan task) error {
...
+
+// AddTaskBalance add task to idle queue
+// do it immediately when no idle queue
+func (p *TaskPool) AddTaskBalance(t task) {
+	length := len(p.qArray)
+
+	// try len/2 times to lookup idle queue
+	for i := 0; i < length/2; i++ {
+		select {
+		case p.qArray[rand.Intn(length)] <- t:
```
Comment (wongoo) [inline sync/task_pool.go]: is it need to check p.done?

**Comment by wongoo** [comment]
new review comment added

File: sync/task_pool.go (hunk 22 above)
Comment (micln) [reply sync/task_pool.go]: the method `xxAlways` means it will called anyway

File: sync/task_pool.go (hunk 23 above)
Comment (micln) [reply sync/task_pool.go]: method `AddTaskBalance` means `AddTaskAlwaysInBalance`

**Comment by wongoo** [comment]
LGTM

**Pull request merged**